### PR TITLE
Sharness Tests

### PR DESCRIPTION
**Pull request by ZenGround0**
Adding sharness tests to test
  -basic commands (version, help, formatting)
  -running cluster-service (init, launch)
  -cluster-ctl commands reporting status of cluster and commands expected to fail
  -cluster-ctl basic pinning and unpinning
  -cleanup to remove config directories and daemon processes

Note -- I am still tweaking the build interface for result aggregation

### CODE REVIEW

**Comment by ZenGround0** [comment]
After the last commit I am tentatively finished, and ready for comments.

File: test/sharness/t0020-service-startup-init.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+test_expect_success "ipfs is installed on this machine" '
+   ipfs help >ipfs_help.txt &&
+   egrep -i "^Usage" ipfs_help.txt >/dev/null  
+'
+
+test_expect_success "launch ipfs daemon" '
+   mkdir -p ../.test_ipfs &&
+   IPFS_PATH="../.test_ipfs" eval '"'"'ipfs init'"'"' &&
+   IPFS_PATH="../.test_ipfs" eval '"'"'ipfs daemon & echo $! >../dPID.txt'"'"' &&
```
Comment (hsanjuan) [inline test/sharness/t0020-service-startup-init.sh]: This evals are very esoteric. Can you try making a separate regular bash function to wrap ipfs launch ipfs and see if the `echo $!` works without evals and stuff there?

Also, store the PIDs in a variable. No need to write files for that.

File: test/sharness/t0020-service-startup-init.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+   sleep 2 
+'
+
+test_expect_success "test config folder exists" '
```
Comment (hsanjuan) [inline test/sharness/t0020-service-startup-init.sh]: Test your prerequisite for ipfs is true. That should imply that it is running with the test folder.

File: test/sharness/t0020-service-startup-init.sh [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+. lib/test-lib.sh
+
+test_expect_success "ipfs is installed on this machine" '
+   ipfs help >ipfs_help.txt &&
+   egrep -i "^Usage" ipfs_help.txt >/dev/null  
+'
+
+test_expect_success "launch ipfs daemon" '
+   mkdir -p ../.test_ipfs &&
+   IPFS_PATH="../.test_ipfs" eval '"'"'ipfs init'"'"' &&
```
Comment (hsanjuan) [inline test/sharness/t0020-service-startup-init.sh]: use a proper tmp path with `mktemp -d` (`mktemp -d ipfs-XXXXXX`). Store name in a variable. Make sure you have a final_cleanup to remove it. You can use `-p` to make the temp directory relative to this folder if you prefer, but don't store stuff in parent directories which don't belong to sharness tests.

File: test/sharness/t0020-service-startup-init.sh [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+   sleep 2 
+'
+
+test_expect_success "test config folder exists" '
+    mkdir -p ../.test_config
+'
+
+test_expect_success "init cluster-service" '
+    ipfs-cluster-service -f --config ../.test_config init 2>service_init.txt &&
+    grep "configuration written" service_init.txt >/dev/null &&
```
Comment (hsanjuan) [inline test/sharness/t0020-service-startup-init.sh]: don't grep output info messages. Test that the actual file has been created. Test that it contains a private key and an ID. (maybe use `jq`).

Also, make a separate test to check that the `-f` does it's job and replaces an existing configuration.

File: test/sharness/t0020-service-startup-init.sh [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    rm service_init.txt
+'
+
+test_expect_success "run cluster-service" '
+    ipfs-cluster-service --config ../.test_config 2>service_start.txt &
+    echo $!>../sPID.txt &&
```
Comment (hsanjuan) [inline test/sharness/t0020-service-startup-init.sh]: store in variable

File: test/sharness/t0020-service-startup-init.sh [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+   IPFS_PATH="../.test_ipfs" eval '"'"'ipfs daemon & echo $! >../dPID.txt'"'"' &&
+   sleep 2 
+'
+
+test_expect_success "test config folder exists" '
+    mkdir -p ../.test_config
+'
+
+test_expect_success "init cluster-service" '
+    ipfs-cluster-service -f --config ../.test_config init 2>service_init.txt &&
```
Comment (hsanjuan) [inline test/sharness/t0020-service-startup-init.sh]: `mktemp -d` to store config location. test_when_finished to cleanup.

File: test/sharness/t0020-service-startup-init.sh [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    grep "configuration written" service_init.txt >/dev/null &&
+    rm service_init.txt
+'
+
+test_expect_success "run cluster-service" '
+    ipfs-cluster-service --config ../.test_config 2>service_start.txt &
```
Comment (hsanjuan) [inline test/sharness/t0020-service-startup-init.sh]: running cluster service should be part of a prerequisite (like ipfs). You can add a test to make sure the pre-requisite exists). Tip: you can put functions which provide common functionality for several test  files in a different file and the source it from every of them.

File: test/sharness/t0020-service-startup-init.sh [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    echo $!>../sPID.txt &&
+    sleep 2  &&
+    egrep -i "ready" service_start.txt >/dev/null &&
```
Comment (hsanjuan) [inline test/sharness/t0020-service-startup-init.sh]: -q

File: test/sharness/t0020-service-startup-init.sh [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    rm service_start.txt
```
Comment (hsanjuan) [inline test/sharness/t0020-service-startup-init.sh]: test_when_finished()

File: test/sharness/t0025-ctl-status-report-commands.sh [hunk 10]
```diff
@@ -0,0 +1,53 @@
+#!/bin/sh
+#
+# MIT Licensed
+#
+
+test_description="Test ctl's status reporting functionality.  Test errors on incomplete commands"
+
+. lib/test-lib.sh
+
+test_expect_success "cluster-ctl version looks good" '
```
Comment (hsanjuan) [inline test/sharness/t0025-ctl-status-report-commands.sh]: remove

File: test/sharness/t0025-ctl-status-report-commands.sh [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    ipfs-cluster-ctl version >version.txt &&
+    egrep "[0-9]+\.[0-9]+\.[0-9]" version.txt >/dev/null &&
+    rm version.txt
+'
+
+test_expect_success "cluster-ctl can read id" '
+    ipfs-cluster-ctl id >id.txt &&
+    grep "> Addresses:" id.txt >/dev/null &&
```
Comment (hsanjuan) [inline test/sharness/t0025-ctl-status-report-commands.sh]: In this test im happy if `ipfs-cluster-ctl id | grep -q "id in the configuration"` succeeds.

File: test/sharness/t0025-ctl-status-report-commands.sh [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    grep "> IPFS: " id.txt >/dev/null &&
+    rm id.txt
+'
+
+test_expect_success "cluster-ctl list 0 peers" '
+    ipfs-cluster-ctl peers ls >peers.txt &&
```
Comment (hsanjuan) [inline test/sharness/t0025-ctl-status-report-commands.sh]: you can check that output from `peers ls` == output from `id`

File: test/sharness/t0025-ctl-status-report-commands.sh [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    grep "| 0 peers" peers.txt >/dev/null &&
+    rm peers.txt
+'
+
+test_expect_failure "cluster-ctl add need multiaddress" '
+    ipfs-cluster-ctl peers add 
+'
```
Comment (hsanjuan) [inline test/sharness/t0025-ctl-status-report-commands.sh]: add tests expecting to fail with invalid multihash

File: test/sharness/t0025-ctl-status-report-commands.sh [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+test_expect_failure "cluster-ctl rm need multihash" '
+    ipfs-cluster-ctl peers rm 
+'
```
Comment (hsanjuan) [inline test/sharness/t0025-ctl-status-report-commands.sh]: add test expecting to fail with invalid peer id

File: test/sharness/t0025-ctl-status-report-commands.sh [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    ipfs-cluster-ctl peers ls >peers.txt &&
+    grep "| 0 peers" peers.txt >/dev/null &&
+    rm peers.txt
+'
+
+test_expect_failure "cluster-ctl add need multiaddress" '
+    ipfs-cluster-ctl peers add 
+'
+
+test_expect_failure "cluster-ctl rm need multihash" '
```
Comment (hsanjuan) [inline test/sharness/t0025-ctl-status-report-commands.sh]: 's/multihash/peer id' for clarity

File: test/sharness/t0100-cleanup.sh [hunk 16]
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
+#
+# MIT Licensed
+#
+
+test_description="Cleanup state from tests (kill ipfs daemon)"
```
Comment (hsanjuan) [inline test/sharness/t0100-cleanup.sh]: cleanups should happen within each sharness test file. This means initialize/kill ipfs and cluster on each test file. Every test file should be self-contained: can be run separately and when it's done there is no trash lying around.

File: test/sharness/t0025-ctl-status-report-commands.sh [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+    ipfs-cluster-ctl peers rm 
+'
+
+test_expect_success "empty cluster-ctl status succeeds" '
```
Comment (hsanjuan) [inline test/sharness/t0025-ctl-status-report-commands.sh]: Need to add a test for `status <cid>`

File: test/sharness/t0025-ctl-status-report-commands.sh [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+test_expect_success "empty cluster-ctl sync succeeds" '
+    ipfs-cluster-ctl sync 
+'
+
+test_expect_failure "empty cluster_ctl recover needs CID" '
+    ipfs-cluster-ctl recover 
+'
+
+test_expect_success "pin ls " '
```
Comment (hsanjuan) [inline test/sharness/t0025-ctl-status-report-commands.sh]: need to add test for `pin ls <cid>`

File: test/sharness/t0030-ctl-pin.sh [hunk 19]
```diff
@@ -0,0 +1,23 @@
+test_description="Test cluster-ctl's pinning and unpinning functionality"
```
Comment (hsanjuan) [inline test/sharness/t0030-ctl-pin.sh]: forgot the shebang?

File: test/sharness/t0030-ctl-pin.sh [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+
+
+test_expect_success "pin data to cluster with ctl" '
+    IPFS_PATH="../.test_ipfs" eval '"'"'ipfs add ../lib/sharness/sharness.sh >add_output.txt'"'"' &&
+    grep "added" add_output.txt &&
+    awk '"'"'{print $2}'"'"' add_output.txt > CID.txt &&
+    ipfs-cluster-ctl pin add $(< CID.txt) >pin_output.txt 
+    grep " PINNED" pin_output.txt >/dev/null &&
+    rm add_output.txt pin_output.txt
+'
```
Comment (hsanjuan) [inline test/sharness/t0030-ctl-pin.sh]: I think this can be expressed in simpler terms:

```
cid=$(echo "test" | ipfs add -q) &&
ipfs-cluster-ctl pin add "$cid" &&
ipfs-cluster-ctl pin ls "$cid" | grep -q "$cid" &&
ipfs-cluster-ctl status "$cid" | grep -q -i "PINNED"
```

**Comment by ZenGround0** [comment]
Thanks for the thorough review!  I'll likely be posting followup questions over the next two days.

**Comment by ZenGround0** [comment]
I have completed a round of fixes addressing the above comments.  Please let me know about any remaining changes you would like to see.

File: sharness/t0020-service-basic-commands.sh [hunk 21]
```diff
@@ -0,0 +1,37 @@
+#!/bin/sh
+
+test_description="Test service startup and init functionality"
+
+. lib/test-lib.sh
+
+test_expect_success "ipfs setup" '
+    test_ipfs_init &&
+    test_have_prereq IPFS_INIT
```
Comment (hsanjuan) [inline sharness/t0020-service-basic-commands.sh]: How I'd do this:

```
. lib/test-lib.sh

test_ipfs_init
test_cluster_init

test_expect_success "prerequisites" '
  test_have_prereq IPFS_INIT
  test_have_prereq CLUSTER_INIT
  ....
  '
```

Then, in every other test, only run if the prerequisite is set:

```
test_expect_success IPFS_INIT,CLUSTER_INIT "ipfs cluster setup" '...'
```

Please do this this way in every test file. Right now tests are running even when prerequisites aren't there so there will be many failures if ipfs fails to start, making it hard to see. It should just fail the prerrequisites test and skip every other test.

File: sharness/lib/test-lib.sh [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+
+# Set prereqs 
+test_ipfs_init() {
+    ipfs help | egrep -q -i "^Usage" &&
+    IPFS_TEMP_DIR=`mktemp -d ipfs-XXXXX` && # Store in TEMP_DIR for safer delete
+    export IPFS_PATH=$IPFS_TEMP_DIR &&
+    ipfs init && 
+    eval 'ipfs daemon & export IPFS_D_PID=`echo $!`' && # Esoteric, but gets correct value of $!
+    sleep 2 &&
+    test_set_prereq IPFS_INIT
```
Comment (hsanjuan) [inline sharness/lib/test-lib.sh]: Probably nicer if we rename the prereq to `IPFS`

File: sharness/lib/test-lib.sh [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+test_cluster_init() {
+    ipfs-cluster-service help | egrep -q -i "^Usage" &&
+    CLUSTER_TEMP_DIR=`mktemp -d cluster-XXXXX` &&
+    ipfs-cluster-service -f --config $CLUSTER_TEMP_DIR init &&
+    eval 'ipfs-cluster-service --config $CLUSTER_TEMP_DIR & export CLUSTER_D_PID=`echo $!`' && 
+    sleep 2 &&
+    test_set_prereq CLUSTER_INIT
```
Comment (hsanjuan) [inline sharness/lib/test-lib.sh]: And this pre-req to CLUSTER

File: Makefile [hunk 24]
```diff
@@ -54,12 +55,24 @@ gx: $(gx_bin) $(gx-go_bin)
...
+	@rm -rf sharness/trash\ directory*
+
 rw: gx
 	$(gx-go_bin) rewrite
 rwundo: gx
 	$(gx-go_bin) rewrite --undo
 publish: rwundo
 	$(gx_bin) publish
-.PHONY: all gx deps test rw rwundo publish service ctl install clean
+.PHONY: all gx deps test test_sharness sharness_deps rw rwundo publish service ctl install clean
```
Comment (hsanjuan) [inline Makefile]: clean_sharness here too

File: Makefile [hunk 25]
```diff
@@ -54,12 +55,24 @@ gx: $(gx_bin) $(gx-go_bin)
 deps: gx
 	$(gx_bin) install --global
 	$(gx-go_bin) rewrite
+
 test: deps
 	go test -tags silent -v ./...
+test_sharness: sharness_deps
+	@sh sharness/run-sharness-tests.sh
+
+sharness_deps: clean_sharness
```
Comment (hsanjuan) [inline Makefile]: like with gx, there is a clever way to make this so that makefile won't re-download sharness if its there.

```
$(sharness):
  steps to download sharness
```
The value of `sharness` should probably be: `sharness/test/sharness`. See my comments on the install file. You can replicate the gx rules pretty much.

File: sharness/t0020-service-basic-commands.sh (hunk 21 above)
Comment (ZenGround0) [inline sharness/t0020-service-basic-commands.sh]: Does this apply to t0010 as well?  I am thinking no because it does not make use of any daemons but perhaps we want consistency.

Also do you have any objections to:

``test_expect_success "ipfs cluster setup" '
        if ! test_have_prereq IPFS; then
            skip_all="..."
            test_done
        fi '
``
or is there a better way to skip all subsequent tests by including the prereqs as the 1st arg to test_expect_success as you wrote above?

**Comment by jbenet** [comment]
@hsanjuan 
> Avoid writing stuff to files. I think you mostly need a temp folder to store ipfs/cluster configuration. Other things can be stored in variables.

curious what is the issue with writing to files? the git sharness tests write most outputs to files and leave the litter, so that you can go in, and check relevant outputs. Example: https://github.com/git/git/blob/master/t/t1000-read-tree-m-3way.sh#L346 -- the sharness scaffolding should make sure to keep the litter self-contained and remove everything before every run.

File: sharness/t0025-ctl-status-report-commands.sh [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+'
+
+test_expect_success JQ "cluster-ctl can read id" '
+    test_cluster_config &&
+    ipfs-cluster-ctl id | egrep -q -i "$CLUSTER_CONFIG_ID"
+'
+
+test_expect_success "cluster-ctl list 0 peers" '
+    export PEER_OUT=`ipfs-cluster-ctl peers ls` &&
+    sorted_peer_out=$(printf '"'"'%s\n'"'"' $PEER_OUT | sort -u) &&
```
Comment (jbenet) [inline sharness/t0025-ctl-status-report-commands.sh]: yeah, dont think that's necessary either.

**Comment by ZenGround0** [comment]
@hsanjuan Another round of fixes.  As usual let me know what needs fixing!

**Comment by hsanjuan** [comment]
@jbenet yeah, fair enough. I dislike tests leaving trash, which means keeping an eye on cleaning what needs to be cleaned, which means extra code when the alternative was just to pipe or to store in a variable. But I realize now there are also good reasons to use files and leave them around when tests fail for inspection (not in all cases perhaps, but in some at least).

File: README.md [hunk 27]
```diff
@@ -34,8 +34,7 @@
   - [Docker](#docker)
   - [Install from sources](#install-from-sources)
 - [Usage](#usage)
-  - [`ipfs-cluster-service`](#ipfs-cluster-service)
-  - [`ipfs-cluster-ctl`](#ipfs-cluster-ctl)
+  - [`Quickstart`](#quickstart)
```
Comment (hsanjuan) [inline README.md]: You did something weird if this is in the diff. You probably need something like:

`git remote add upstream https://github.com/ipfs/ipfs-cluster.git`
`git fetch upstream`
`git rebase upstream/master`

and fix conflicts along the way:

`fix conflict`
`git add <problem file>`
`git rebase --continue`

at the end

`git push origin <your branch> --force`

so that only your changes are in this PR.

**Comment by ZenGround0** [comment]
@hsanjuan Alright this should do it!   I tried to rebase before the last push which is where the weirdness came from I think, but I think the history is clean now.  The trailing whitespace should be gone from all files I touched and I fixed the problems you pointed out in the Makefile changes.

**Comment by hsanjuan** [comment]
Thanks!! Next step (in a new PR) is to fix broken tests and enable them in travis.
